dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/DefaultAnalyticsServiceTest.java

```diff
@@ -0,0 +1,127 @@
+ */
+package org.hisp.dhis.analytics.data;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.hisp.dhis.common.BaseDimensionalItemObject;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.common.Grid;
+import org.hisp.dhis.common.GridHeader;
+import org.junit.jupiter.api.Test;
+
+class DefaultAnalyticsServiceTest {
+
+  @Test
+  void testGetGridItems() {
+
+    Grid grid = mock(Grid.class);
+
+    when(grid.getWidth()).thenReturn(6);
+    when(grid.getRows())
+        .thenReturn(
+            List.of(
+                List.of("R-A1", "R-B1", "R-C2", "C-A1", "C-B1", "v1"),
+                List.of("R-A1", "R-B2", "R-C1", "C-A1", "C-B3", "v2"),
+                List.of("R-A2", "R-B1", "R-C1", "C-A2", "C-B1", "v3")));
+    when(grid.getHeaders())
+        .thenReturn(
+            List.of(
+                new GridHeader("R-A", false, false),
+                new GridHeader("R-B", false, false),
+                new GridHeader("R-C", false, false),
+                new GridHeader("C-A", false, false),
+                new GridHeader("C-B", false, false)));
+
+    Map<String, List<DimensionalItemObject>> dimensionItemsByDimension =
+        Map.of(
+            "R-A",
+                List.of(
+                    new BaseDimensionalItemObject("R-A1"),
+                    new BaseDimensionalItemObject("R-A2"),
+                    new BaseDimensionalItemObject("R-A3")),
+            "R-B",
+                List.of(
+                    new BaseDimensionalItemObject("R-B1"),
+                    new BaseDimensionalItemObject("R-B2"),
+                    new BaseDimensionalItemObject("R-B3")),
+            "R-C",
+                List.of(
+                    new BaseDimensionalItemObject("R-C1"),
+                    new BaseDimensionalItemObject("R-C2"),
+                    new BaseDimensionalItemObject("R-C3")),
+            "C-A",
+                List.of(
+                    new BaseDimensionalItemObject("C-A1"),
+                    new BaseDimensionalItemObject("C-A2"),
+                    new BaseDimensionalItemObject("C-A3")),
+            "C-B",
+                List.of(
+                    new BaseDimensionalItemObject("C-B1"),
+                    new BaseDimensionalItemObject("C-B2"),
+                    new BaseDimensionalItemObject("C-B3")));
+
+    List<String> dimensionIds = List.of("R-A", "R-B", "R-C", "C-A", "C-B");
+
+    List<List<DimensionalItemObject>> gridItems =
+        DefaultAnalyticsService.getGridItems(grid, dimensionItemsByDimension, dimensionIds);
+
+    assertEquals(grid.getRows().size(), gridItems.size());
+
+    for (int i = 0; i < grid.getRows().size(); i++) {
+      assertEquals(grid.getWidth() - 1, gridItems.get(i).size());
+    }
+
+    Set<String> fromGrid =
```

Comment:
Could we simplify it? This is the kind of code we would like to avoid, as discussed days ago...